crates/red_knot_python_semantic/src/types/class.rs

```diff
@@ -290,9 +290,23 @@ impl<'db> ClassType<'db> {
     /// traverse through the MRO until it finds the member.
     pub(super) fn own_class_member(self, db: &'db dyn Db, name: &str) -> SymbolAndQualifiers<'db> {
         let (class_literal, _) = self.class_literal(db);
-        class_literal
-            .own_class_member(db, name)
-            .map_type(|ty| self.specialize_type(db, ty))
+        class_literal.own_class_member(db, name).map_type(|ty| {
+            let specialized = self.specialize_type(db, ty);
+            // The `__new__` and `__init__` members of a non-specialized generic class are handled
+            // specially: they inherit the generic context of their class. That lets us treat them
+            // as generic functions when constructing the class, and infer the specialization of
+            // the class from the arguments that are passed in.
```

Comment:
This is where we might consider having _all_ class methods inherit the class's generic context, if we want to e.g. support inferring a specialization for a factory method.

Reply:
I think this could be a comment in the code, for better access later.

Reply:
Done